SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp

```diff
@@ -0,0 +1,550 @@
+	}
+
+	return ProjectName;
+}
+
+void FLocalDeploymentManager::WorkerBuildConfigAsync()
+{
+	AsyncTask(ENamedThreads::AnyBackgroundThreadNormalTask, [this]
+	{
+		FString SpatialExe = TEXT("spatial.exe");
+		FString SpatialServiceStatusArgs = TEXT("worker build build-config");
+		const FString SpatialDirectory = GetSpatialOSDirectory();
+
+		FString WorkerBuildConfigResult;
+		int32 ExitCode;
+		ExecuteAndReadOutput(*SpatialExe, *SpatialServiceStatusArgs, *SpatialDirectory, WorkerBuildConfigResult, ExitCode);
+
+		if (ExitCode == ExitCodeSuccess)
+		{
+			UE_LOG(LogSpatialDeploymentManager, Error, TEXT("Building worker configurations failed. Please ensure your .worker.json files are correct."));
+		}
+	});
+}
+
+bool FLocalDeploymentManager::ParseJson(FString RawJsonString, TSharedPtr<FJsonObject>& JsonParsed)
+{
+	TSharedRef<TJsonReader<TCHAR>> JsonReader = TJsonReaderFactory<TCHAR>::Create(RawJsonString);
+	return FJsonSerializer::Deserialize(JsonReader, JsonParsed);
+}
+
+// ExecuteAndReadOutput exists so that a spatial command window does not spawn when using 'spatial.exe'. It does not however allow reading from StdErr.
+// For other processes which do not spawn cmd windows, use ExecProcess instead.
+void FLocalDeploymentManager::ExecuteAndReadOutput(FString Executable, FString Arguments, FString DirectoryToRun, FString& OutResult, int32& ExitCode)
+{
+	UE_LOG(LogSpatialDeploymentManager, Verbose, TEXT("Executing '%s' with arguments '%s' in directory '%s'"), *Executable, *Arguments, *DirectoryToRun);
+
+	void* ReadPipe = nullptr;
+	void* WritePipe = nullptr;
+	ensure(FPlatformProcess::CreatePipe(ReadPipe, WritePipe));
+
+	FProcHandle ProcHandle = FPlatformProcess::CreateProc(*Executable, *Arguments, false, true, true, nullptr, 1 /*PriorityModifer*/, *DirectoryToRun, WritePipe);
+
+	if (ProcHandle.IsValid())
+	{
+		for (bool bProcessFinished = false; !bProcessFinished; )
+		{
+			bProcessFinished = FPlatformProcess::GetProcReturnCode(ProcHandle, &ExitCode);
+
+			OutResult = OutResult.Append(FPlatformProcess::ReadPipe(ReadPipe));
+			FPlatformProcess::Sleep(0.01f);
+		}
+
+		ProcHandle.Reset();
+	}
+	else
+	{
+		UE_LOG(LogSpatialDeploymentManager, Error, TEXT("Execution failed. '%s' with arguments '%s' in directory '%s' "), *Executable, *Arguments, *DirectoryToRun);
+	}
+
+	FPlatformProcess::ClosePipe(0, ReadPipe);
+	FPlatformProcess::ClosePipe(0, WritePipe);
+}
+
+void FLocalDeploymentManager::RefreshServiceStatus()
+{
+	FDateTime CurrentTime = FDateTime::Now();
+
+	FTimespan Span = CurrentTime - LastSpatialServiceCheck;
+
+	if (Span.GetSeconds() < RefreshFrequency)
+	{
+		return;
+	}
+
+	AsyncTask(ENamedThreads::AnyBackgroundThreadNormalTask, [this]
+	{
+		GetServiceStatus();
+		GetLocalDeploymentStatus();
+	});
+
+	LastSpatialServiceCheck = FDateTime::Now();
+}
+
+bool FLocalDeploymentManager::TryStartLocalDeployment(FString LaunchConfig, FString LaunchArgs)
```

Comment:
Do the strings need to be copied?

Reply:
Because the calls to TryStartLocalDeployment are within Async tasks I'm going to say yes, we don't know when the strings could be destroyed.